vumi_message_store/memory_backend_manager.py

```diff
@@ -0,0 +1,371 @@
+            "content_type": fake_object._current_data["content_type"],
+            "data": json.dumps(fake_object._current_data["data"]),
+        }
+        indexes = {}
+        for index_name, index_value in fake_object._current_data["indexes"]:
+            indexes.setdefault(index_name, []).append(
+                (index_value, fake_object.key))
+        for index_name, index_values in indexes.iteritems():
+            old_indexes = bucket["indexes"].get(index_name, [])
+            new_indexes = [
+                (value, key) for value, key in old_indexes
+                if key != fake_object.key]
+            new_indexes.extend(index_values)
+            new_indexes.sort()
+            bucket["indexes"][index_name] = new_indexes
+
+    def _reload_object_state(self, fake_object):
+        """
+        Set an object's data from stored state.
+        """
+        bucket = self._get_bucket(fake_object._bucket_name)
+        obj_state = bucket["objects"].get(fake_object.key)
+        if obj_state is None:
+            fake_object._current_data["data"] = None
+            return
+        fake_object._current_data["data"] = json.loads(obj_state["data"])
+        fake_object._current_data["content_type"] = obj_state["content_type"]
+        fake_object._current_data["indexes"] = set()
+        for index_name, index_data in bucket["indexes"].iteritems():
+            for index_value, key in index_data:
+                if key == fake_object.key:
+                    fake_object._current_data["indexes"].add(
+                        (index_name, index_value))
+
+    def get_index_page(self, bucket_name, index_name, start_value, end_value,
+                       return_terms=False, max_results=None,
+                       continuation=None):
+        """
+        Perform an index query.
+        """
+        # Get all matching results
+        index = self._get_bucket(bucket_name)["indexes"].get(index_name, [])
+        if end_value is None:
+            in_range = lambda v: v[0] == start_value
+        else:
+            in_range = lambda v: start_value <= v[0] <= end_value
+        results = filter(in_range, index)
+
+        # Drop all results we've returned previously, if any.
+        if continuation is not None:
+            continuation = tuple(json.loads(base64.b64decode(continuation)))
+            while results and results[0] < continuation:
+                results.pop(0)
+
+        # If we're not paginated, we're done.
+        if max_results is None:
+            return self._return_index_page(
+                bucket_name, index_name, start_value, end_value, return_terms,
+                max_results, results, continuation=None)
+
+        # Truncate the results and build the continuation token.
+        continuation = None
+        truncated_results = results[:max_results]
+        if len(truncated_results) < len(results):
+            continuation = base64.b64encode(
+                json.dumps(list(results[max_results])))
```

Comment:
Probably left over from an earlier implementation attempt that used iterators or something. I'll remove it.